---
title: "Canonical correlation analysis"
subtitle: ''
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
documentclass: article
output: 
  html_document:
  toc: true
  smart: false
vignette: >
  %\VignetteIndexEntry{Canonical correlation analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\usepackage[utf8]{inputenc}
---


<!--- 

cd build2/decorrelate/vignettes/


rmarkdown::render("cca.Rmd");

--->



```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  eval = TRUE,
  dev = c("png", "pdf"),
  cache = TRUE)
```

```{r load.packages}
library(decorrelate)
library(Rfast)
library(ggplot2)
library(data.table)
library(CCA)
library(RCCA)
```

Avoids 2d parameter search of https://www.jstatsoft.org/article/view/v023i12


```{r fastcca.scaling}
n = 500
p.array = seq(10, 10000, by=1000)

df = lapply(p.array, function(p){
  message("\r", p, '   ', appendLF=FALSE)

  X = matrnorm(n,p)
  Y = matrnorm(n,p)

  res1 = system.time(decorrelate::cca(X,Y))
  res4 = system.time(decorrelate::fastcca(X,Y))

  rbind(data.frame(Method = "decorrelate::cca", p = p, t(c(res1)), check.names=FALSE),
    data.frame(Method = "fastcca", p = p, t(c(res4))))
  })
df = do.call(rbind, df)

ggplot(df, aes(p, elapsed, color=Method)) + geom_line() + geom_point() + theme_classic()
```



```{r compare.all}
n = 500

p.array = c(seq(10, 1000, by=75), seq(1000, 10000, by=400))

df = lapply(p.array, function(p){

  message("\r", p, "   ", appendLF=FALSE)

  X = matrnorm(n,p)
  Y = matrnorm(n,p)

  res = c()

  if( p < 9000){
    tm = system.time(decorrelate::cca(X,Y))
    res = rbind(res, data.frame(Method = "decorrelate::cca", p = p, t(c(tm)), check.names=FALSE))
  }

  if( p < Inf){
    tm = system.time(RCCA(X,Y, 0.1, 0.1))
    res = rbind(res, data.frame(Method = "RCCA::RCCA", p = p, t(c(tm)), check.names=FALSE))
  }

  if( p < 800){
    tm = system.time(rcc(X,Y, 0.1, 0.1))
    res = rbind(res, data.frame(Method = "CCA::rcc", p = p, t(c(tm)), check.names=FALSE))
  }

  if( p < 5000){
    tm = system.time(whitening::scca(X,Y, verbose=FALSE))
    res = rbind(res, data.frame(Method = "whitening::scca", p = p, t(c(tm)), check.names=FALSE))
  }

  if( p < Inf){
    tm = system.time(decorrelate::fastcca(X,Y))
    res = rbind(res, data.frame(Method = "decorrelate::fastcca", p = p, t(c(tm)), check.names=FALSE))
  }

  res
  })
df = do.call(rbind, df)

```{r plot.compare.p, cache=FALSE}
ggplot(df, aes(p, elapsed, color=Method)) + geom_line() + geom_point() + theme_classic() + ylab("Wall time (seconds)") + scale_color_brewer(palette="Set1") + theme(aspect.ratio=1)
```


```{r compare.all.fixed.p}
p = 1000

n.array = c(seq(10, 1000, by=75), seq(1000, 10000, by=400))

df = lapply(n.array, function(n){

  message("\r", n, "   ", appendLF=FALSE)

  X = matrnorm(n,p)
  Y = matrnorm(n,p)

  res = c()

  if( n < Inf){
    tm = system.time(decorrelate::cca(X,Y))
    res = rbind(res, data.frame(Method = "decorrelate::cca", n=n, t(c(tm)), check.names=FALSE))
  }

  if( n < Inf){
    tm = system.time(RCCA::RCCA(X,Y, 0.1, 0.1))
    res = rbind(res, data.frame(Method = "RCCA::RCCA", n=n, t(c(tm)), check.names=FALSE))
  }

  # if( n < 1000){
  #   tm = system.time(CCA::rcc(X,Y, 0.1, 0.1))
  #   res = rbind(res, data.frame(Method = "CCA::rcc", n=n, t(c(tm)), check.names=FALSE))
  # }

  if( n < Inf){
    tm = system.time(whitening::scca(X,Y, verbose=FALSE))
    res = rbind(res, data.frame(Method = "whitening::scca", n=n, t(c(tm)), check.names=FALSE))
  }

  if( n < Inf){
    tm = system.time(decorrelate::fastcca(X,Y))
    res = rbind(res, data.frame(Method = "decorrelate::fastcca", n=n, t(c(tm)), check.names=FALSE))
  }

  res
  })
df = do.call(rbind, df)

ggplot(df, aes(n, elapsed, color=Method)) + geom_line() + geom_point() + theme_classic() + ylab("Wall time (seconds)") + scale_color_brewer(palette="Set1") + theme(aspect.ratio=1)
```




```{r test.sims}
# Test sims
############

# rmarkdown::render("cca.Rmd")


# devtools::reload("/Users/gabrielhoffman/workspace/repos/decorrelate")


library(decorrelate)
library(Rfast)
library(ggplot2)
library(Matrix)
library(Matrix)

n = 400
p = 30
q = 30

sig.array = seq(.05, .99, length.out=50)

sig_within = .7

df_cca = lapply( sig.array, function(sig_within){

  message("\r", format(sig_within, digits=3), appendLF = FALSE)

  Sigma.xx = diag(1,p)
  Sigma.xx[Sigma.xx!=1] = sig_within
  Sigma.yy = diag(1,q)
  Sigma.yy[Sigma.yy!=1] = sig_within

  Sigma.xy = matrix(sig_within^2, p, q)

  Sig = rbind(cbind(Sigma.xx, Sigma.xy), cbind(t(Sigma.xy), Sigma.yy))

  dat = Rfast:::rmvnorm(n, rep(0, p+q), Sig, seed=1)

  # correlation between samples 
  S = bdiag(lapply(seq(1,n / 100), function(i) autocorr.mat(100, .95)))
  dat = chol(S) %*% dat 

  X = scale(dat[,1:p])
  Y = scale(dat[,-c(1:p)]) 
  
  incl = seq(1, 0.8*n)   

  fit.fastcca = decorrelate::fastcca(X[incl,],Y[incl,])
  X.pred = X[-incl,] %*% fit.fastcca$x.coef
  Y.pred = Y[-incl,] %*% fit.fastcca$y.coef
  rho.fastcca = abs(cor(X.pred[,1], Y.pred[,1]))

  fit.d.cca = decorrelate::cca(X[incl,],Y[incl,])
  X.pred = X[-incl,] %*% fit.d.cca$x.coefs
  Y.pred = Y[-incl,] %*% fit.d.cca$y.coefs
  rho.d.cca = abs(cor(X.pred[,1], Y.pred[,1]))

  fit.scca = whitening:::scca( X[incl,], Y[incl,], verbose=FALSE)
  X.pred = X[-incl,] %*% t(fit.scca$WX)
  Y.pred = Y[-incl,] %*% t(fit.scca$WY)
  rho.scca = abs(cor(X.pred[,1], Y.pred[,1]))

  fit.rcc = CCA::rcc(X[incl,], Y[incl,], .1, .1)
  X.pred = X[-incl,] %*% fit.rcc$xcoef
  Y.pred = Y[-incl,] %*% fit.rcc$ycoef
  rho.rcc = abs(cor(X.pred[,1], Y.pred[,1]))

  fit.rcca = RCCA::RCCA(X[incl,], Y[incl,], .1, .1)
  X.pred = X[-incl,] %*% fit.rcca$x.coefs
  Y.pred = Y[-incl,] %*% fit.rcca$y.coefs
  rho.rcca = abs(cor(X.pred[,1], Y.pred[,1]))

  fit.indep.cca = decorrelate::fastcca(X[incl,],Y[incl,], lambda.x=1, lambda.y=1)
  X.pred = X[-incl,] %*% fit.indep.cca$x.coef
  Y.pred = Y[-incl,] %*% fit.indep.cca$y.coef
  rho.indep.cca = abs(cor(X.pred[,1], Y.pred[,1]))

  data.frame(sig_within = sig_within,  
    FastCCA = rho.fastcca, 
   'decorr::CCA' = rho.d.cca,
    'whitening::scca' = rho.scca,
    'CCA::rcc' = rho.rcc,
    "RCCA::RCCA" = rho.rcca,
    "Indep CCA" = rho.indep.cca,
    check.names=FALSE)
})
df_cca = do.call(rbind, df_cca)

df = reshape2::melt(df_cca, id.vars = "sig_within")
df = df[order(df$sig_within),]

ggplot(df, aes(sig_within, value, color=variable)) + geom_point() + scale_color_brewer(palette="Set1") + theme_classic() + theme(aspect.ratio=1) + xlab("Method") + ylab("Correlation of prediction") + geom_smooth() + scale_x_continuous(limits=c(0, 1)) + scale_y_continuous(limits=c(0, 1))
```


```{r plot2, cache=FALSE}
ggplot(df, aes(sig_within, value, color=variable)) + scale_color_brewer(palette="Set1") + theme_classic() + theme(aspect.ratio=1) + ylab("Correlation of prediction") + geom_smooth() + scale_x_continuous(limits=c(0, 1)) + scale_y_continuous(limits=c(0, 1))
```

```{r barplot, cache=FALSE}
df = data.table(df)

df_mean = df[,mean(value), by="variable"]
ymax = 1.03*max(df_mean$V1)
ggplot(df_mean, aes(variable, V1, fill=variable)) + geom_bar(stat="identity") + theme_classic() + theme(aspect.ratio=1) + ylab("Average prediction correlation") + coord_flip() + scale_fill_brewer(palette="Set1") + scale_y_continuous(limits=c(0, ymax), expand=c(0,0))
```


```{r wrap, cache=FALSE}
ggplot(df, aes(sig_within, value, color=variable)) + geom_point() + scale_color_brewer(palette="Set1") + theme_classic() + theme(aspect.ratio=1) + xlab("Method") + ylab("Correlation of prediction") + geom_smooth(color="black") + scale_x_continuous(limits=c(0, 1)) + scale_y_continuous(limits=c(0, 1)) + facet_wrap(~variable)
```

```{r exit, cache=FALSE, echo=FALSE}
knitr::knit_exit()
```

 devtools::reload("/Users/gabrielhoffman/workspace/repos/decorrelate")


system.time(fit.fastcca <- decorrelate::fastcca(X[incl,],Y[incl,]) )
X.pred = X[-incl,] %*% fit.fastcca$x.coef
Y.pred = Y[-incl,] %*% fit.fastcca$y.coef
abs(cor(X.pred[,1], Y.pred[,1]))


source("../R/fastccahd.R")
system.time(fit <- fastccahd(X[incl,],Y[incl,], k.svd=29, std=FALSE))
X.pred = X[-incl,] %*% fit$x.coef
Y.pred = Y[-incl,] %*% fit$y.coef
abs(cor(X.pred[,1], Y.pred[,1]))




source("../R/fastccahd.R")
fit <- fastccahd(X[incl,],Y[incl,], k.svd=30, std=FALSE)
X.pred = X[-incl,] %*% fit$x.coef
Y.pred = Y[-incl,] %*% fit$y.coef
abs(cor(X.pred[,1], Y.pred[,1]))

fit <- fastccahd(X[incl,],Y[incl,], k.svd=29, std=FALSE)
X.pred = X[-incl,] %*% fit$x.coef
Y.pred = Y[-incl,] %*% fit$y.coef
abs(cor(X.pred[,1], Y.pred[,1]))










whitening:::scca preforms better, but its not clear why
Need to step through to see differences in code

l = 0
fit.fastcca = decorrelate::cca(X[incl,],Y[incl,], k=2, l, l)
X.pred = X[-incl,] %*% fit.fastcca$x.coef
Y.pred = Y[-incl,] %*% fit.fastcca$y.coef
cor(X.pred[,1], Y.pred[,1])

fit.scca = whitening:::scca( X[incl,], Y[incl,], lambda.cor=0)
X.pred = X[-incl,] %*% t(fit.scca$PhiX)
Y.pred = Y[-incl,] %*% t(fit.scca$PhiY)
cor(X.pred[,1], Y.pred[,1])


fit.cancor = CCA::cc(X[incl,], Y[incl,])
X.pred = X[-incl,] %*% fit.cancor$xcoef
Y.pred = Y[-incl,] %*% fit.cancor$ycoef
cor(X.pred[,1], Y.pred[,1])


minvsqrt(cor(X,X)) %*% cor(X,Y) %*% minvsqrt(cor(Y,Y))



l.grid = seq(1e-4, 1-1e-4, length.out=30)
lambda.grid = expand.grid(l.grid, l.grid)

df = apply( lambda.grid, 1, function(l){
  fit = decorrelate::fastcca(X[incl,],Y[incl,], k=2,l[1],l[2])
  X.pred = X[-incl,] %*% fit$x.coef
  Y.pred = Y[-incl,] %*% fit$y.coef
  rho = cor(X.pred[,1], Y.pred[,1])
  data.frame(lambda.x = l[1], lambda.y = l[2], rho=rho)
  })
df = do.call(rbind, df)

ggplot(df, aes(lambda.x, lambda.y, z=rho)) + geom_contour_filled() + theme_minimal() + theme(aspect.ratio=1)




# Test predictions
```{r test.pred}

n = 1000
p = 400

Sigma = autocorr.mat(p, .9)

k = 1
X_latent = cbind(c(rep(1,n/2), rep(0,n/2)), c(rep(.6,n/2), rep(1,n/2)))


alpha = matrnorm(2,1)
mu.x = X_latent %*% alpha
X = mu.x + rmvnorm(n, rep(0, p), Sigma)



beta = rnorm(2)
mu.y = X_latent %*% beta
Y = t(rmvnorm(n, mu.y, Sigma))

incl = sample.int(n, 80, replace=FALSE)

l = 1e-5
fit = decorrelate::fastcca(X[incl,],Y[incl,], k=2,l,l)

X.pred = X[-incl,] %*% fit$x.coef
Y.pred = Y[-incl,] %*% fit$y.coef


a = cor(X.pred[,1], X_latent[-incl,1])
b = cor(Y.pred[,1], X_latent[-incl,1])

abs(a) + abs(b)


par(mfrow=c(3,2))
plot(X.pred)
plot(Y.pred)
plot(X.pred[,1], Y.pred[,1])
plot(X.pred[,2], Y.pred[,2])



library(whitening)
data(nutrimouse)

l = NULL
fit = fastcca(scale(nutrimouse$gene), scale(nutrimouse$lipid), lambda.x=l, lambda.y=l)

par(mfrow=c(3,2))
plot(fit$x.vars[,1], fit$y.vars[,1], col=nutrimouse$genotype )
plot(fit$x.vars[,2], fit$y.vars[,2], col=nutrimouse$diet )
plot(fit$x.vars[,3], fit$y.vars[,3], col=nutrimouse$diet )
plot(fit$x.vars[,4], fit$y.vars[,4], col=nutrimouse$diet )


plot(fit$x.vars[,1], fit$x.vars[,2], pch=c(1,2)[nutrimouse$genotype], col=nutrimouse$diet )
plot(fit$y.vars[,1], fit$y.vars[,2], pch=c(1,2)[nutrimouse$genotype], col=nutrimouse$diet )


cca.out = whitening::scca( as.matrix(nutrimouse$gene), as.matrix(nutrimouse$lipid))

CCAX = tcrossprod( scale(X), cca.out$WX )
CCAY = -tcrossprod( scale(Y), cca.out$WY )


par(mfrow=c(3,2))
plot(CCAX[,1], CCAY[,1], col=nutrimouse$genotype )
plot(CCAX[,2], CCAY[,2], col=nutrimouse$diet )
plot(CCAX[,3], CCAY[,3], col=nutrimouse$diet )
plot(CCAX[,4], CCAY[,4], col=nutrimouse$diet )

plot(CCAX[,1], CCAX[,2], pch=c(1,2)[nutrimouse$genotype], col=nutrimouse$diet )
plot(CCAY[,1], CCAY[,2], pch=c(1,2)[nutrimouse$genotype], col=nutrimouse$diet )


dat1 = lol.sims.cigar(100, n3)





C.cca = solve(chol(Sigma.xx)) %*% Sigma.xy %*% solve(chol(Sigma.yy)) 

dcmp.true = svd(C.cca)
dcmp.true$d[1]

cancor(X,Y)$cor[1]




# https://rdrr.io/cran/PMA/man/CCA.html

n = 10000
p = 23
q = 2

u <- scale(cbind(rbinom(n, 3, prob=.5 )))
# v1 = matrnorm(p,1)
# v2 = matrnorm(q,1)
v1 = rep(1,p)
v2 = rep(1,q)

s = 2
X <- tcrossprod(u,v1) + matrnorm(n,p)*sqrt(s)
Y <- tcrossprod(u,v2) + matrnorm(n,q)*sqrt(s)

a = diag(cor(X, tcrossprod(u,v1)))^2
b = diag(cor(Y, tcrossprod(u,v2)))^2

f1 = sum(a) / (sum(a) + s)
f2 = sum(b) / (sum(b) + s)

f1*f2

(sum(a) + sum(b)) / (p+q)



signal = p / 2

fit = cancor(X, Y)
fit$cor[1]^2





# make sure single values don't exceed 1
fit = decorrelate::fastcca(X,Y)
fit$cor[1] 

fit = RCCA::RCCA(X,Y, .1, .1)
fit$mod.cors[1]

fit2 = whitening::scca(X,Y)
fit2$lambda[1]













https://onlinelibrary.wiley.com/doi/epdf/10.1002/hbm.25090
https://www.jstatsoft.org/article/view/v023i12
https://www.embopress.org/doi/full/10.15252/msb.20178124
10.1093/bioinformatics/btaa530
https://www.biorxiv.org/content/10.1101/2020.08.25.265546v3.full.pdf







